pallene/to_ir.lua

```diff
@@ -209,66 +209,123 @@ function ToIR:convert_stat(cmds, stat)
         local decls = stat.decls
         local exps = stat.exps
 
-        -- for-in loops are desugared into regurlar loops before compiling.
-        -- For example, a loop like this:
-        --- ```
-        -- for a: T1, b: T2 in <RHS> do
-        --     <loop body>
-        -- end
-        ---```
-        -- is compiled as if the following was written instead:
-        --```
-        -- local iter, st, a_any, b_any
-        -- iter, st, ctrl = RHS[1], RHS[2], RHS[3]
-        -- while true do
-        --   a_any, b_any = iter(st, ctrl)
-        --   if a_any == nil then break end
-        --   local a = a_any as T1
-        --   local b = b_any as T2
-        --   <loop body>
-        -- end
-        -- ```
-
-        local v_iter = ir.add_local(self.func, "$iter", exps[1]._type)
-        self:exp_to_assignment(cmds, v_iter, exps[1])
-        local v_state = ir.add_local(self.func, "$st", exps[2]._type)
-        self:exp_to_assignment(cmds, v_state, exps[2])
-
-        local v_lhs_dyn = {}
-        for _, decl in ipairs(decls) do
-            local v = ir.add_local(self.func, "$" .. decl.name .. "_dyn", types.T.Any())
-            table.insert(v_lhs_dyn, v)
-        end
 
-        local v_ctrl = v_lhs_dyn[1]
-        self:exp_to_assignment(cmds, v_ctrl, exps[3])
+        if stat.is_ipairs then
+            -- `ipairs` are desugared down to regular for-loops
```

Comment:
I suspect that this desugaring might be wrong in situations where the table has "holes" with nil. When we use ipairs it always stops at the first nil, which might not be the same thing as the `#xs` in a table with holes.

What do you think about using the same desugaring that we used for the regular for-in loop, except that instead of calling an iterator function, we use `xs[i]` and `i = i+1` instead? Basically just "inlining" the definition of the iter function for ipairs.